File: tag_detector/tag_detector.py
import numpy as np
import cv2
import copy
import time

def contour_generator(frame):
    test_img1 = cv2.cvtColor(frame, cv2.COLOR_BGR2GRAY)
    test_blur = cv2.GaussianBlur(test_img1, (7, 7), 0)
    edge = cv2.Canny(test_blur, 75, 175)
    edge1 = copy.copy(edge)
    contour_list = list()

    cnts, h = cv2.findContours(edge1, cv2.RETR_TREE, cv2.CHAIN_APPROX_SIMPLE)
    
    # error handling
    
    if h is None:
        return []

    index = list()
    for hier in h[0]:
        if hier[3] != -1:
            index.append(hier[3])

    # loop over the contours
    for c in index:
        peri = cv2.arcLength(cnts[c], True)
        approx = cv2.approxPolyDP(cnts[c], 0.02 * peri, True)

        if len(approx) > 4:
            peri1 = cv2.arcLength(cnts[c - 1], True)
            corners = cv2.approxPolyDP(cnts[c - 1], 0.02 * peri1, True)
            contour_list.append(corners)

    new_contour_list = list()
    for contour in contour_list:
        if len(contour) == 4:
            new_contour_list.append(contour)
    final_contour_list = list()
    for element in new_contour_list:
        if cv2.contourArea(element) > 700 and cv2.isContourConvex(element):
            final_contour_list.append(element)

    return final_contour_list

# Function to return the order of points in camera frame
def order(pts):
    rect = np.zeros((4, 2), dtype="float32")

    s = pts.sum(axis=1)
    # print(np.argmax(s))
    rect[0] = pts[np.argmin(s)]
    rect[2] = pts[np.argmax(s)]

    diff = np.diff(pts, axis=1)
    # print(np.argmax(diff))
    rect[1] = pts[np.argmin(diff)]
    rect[3] = pts[np.argmax(diff)]

    # return the ordered coordinates
    return rect

lowerRed1 = np.array([0, 50, 50])
upperRed1 = np.array([10, 255, 255])
lowerRed2 = np.array([170, 50, 50])
upperRed2 = np.array([180, 255, 255])

lowerBlue = np.array([100,80,160])
upperBlue = np.array([120,255,255])

def determineColor(marker):
    markerHSV = cv2.cvtColor(marker, cv2.COLOR_BGR2HSV)

    redMask1 = cv2.inRange(markerHSV, lowerRed1, upperRed1)
    redMask2 = cv2.inRange(markerHSV, lowerRed2, upperRed2)
    redMask = redMask1 + redMask2

    blueMask = cv2.inRange(markerHSV, lowerBlue, upperBlue)

    cv2.imshow("redMask1", redMask1)
    cv2.imshow("redMask2", redMask1)
    cv2.imshow("redMask", redMask)
    cv2.imshow("blueMask", blueMask)

    if np.sum(redMask) > 0:
        return "red"
    elif np.sum(blueMask) > 0:
        return "blue"
    return "unknown"

def determineLetter(marker):
    ret, img_bw = cv2.threshold(marker, 200, 255, cv2.THRESH_BINARY)
    white = 255
    cropped_img = img_bw[25:150, 25:150]
    cv2.imshow("Cropped", cropped_img)

    # [y, x]: [0, 0] = top left pixel
    if cropped_img[12, 37] != white:    # Center of pixel at [0, 1]; not white if letter "A"
        return "A"
    elif cropped_img[12, 12] != white:  # Center of pixel at [0, 0]; not white if letter "B"
        return "C"
    elif cropped_img[12, 112] == white: # Center of pixel at [0, 4]; white if letter "E"
        return "E"
    elif cropped_img[62, 112] == white: # Center of pixel at [2, 4]; white if letter "D"
        return "D"
    else:
        return "B"                      # Must be "B" if none of the above



fid_size = 0.053  # meters
# TODO fix calibration to setup cameraMatrix
cameraMatrix = np.array([[1.25649815e+03, 0.0, 7.12996774e+02],
                         [0.0, 1.25820533e+03, 4.69551858e+02], 
                         [0.0, 0.0, 1.0]], dtype=np.float32)
distCoeffs = np.array([[-3.72271817e-03, 5.33786890e-01, -4.99625728e-04, -1.65101232e-03,-1.78505927e+00]], dtype=np.float32)
# TODO testing program with hardcoded matrix since calibration session is not working
# TODO when game ready, dont forget to comment this out

def findTranslationAndRotation(h):
    object_points = np.array([[-fid_size / 2.0, fid_size / 2.0, 0.0],
                              [fid_size / 2.0, fid_size / 2.0, 0.0],
                              [fid_size / 2.0, -fid_size / 2.0, 0.0],
                              [-fid_size / 2.0, -fid_size / 2.0, 0.0]], dtype=np.float32)
    # TODO error here since cameraMatrix is null
    _, rvec, tvec = cv2.solvePnP(object_points, np.array(h, dtype=np.float32), cameraMatrix, distCoeffs, flags=cv2.SOLVEPNP_IPPE_SQUARE)

    if rvec is not None:
        Rt = np.matrix(cv2.Rodrigues(rvec)[0])
        yaw, pitch, roll = rotToEul(Rt)
        R = Rt.T
        pose = -R * np.matrix(tvec)

        print("CAMERA Position x=" + str(pose[2]) + " y=" + str(pose[0]) + " z=" + str(pose[1]))
        print("Camera tilt -------   yaw=" + str(yaw) + " pitch=" + str(pitch) + " roll=" + str(roll))
        return pose, (yaw, pitch, roll)
    
    return "ERROR: Rotation Vector Not Found"

def rotToEul(R):
    sy = np.sqrt(R[0, 0] ** 2 + R[1, 0] ** 2)

    singular = sy < 1e-6
    if not singular:
        yaw = np.arctan2(R[2, 1], R[2, 2])  # Rotation around Z-axis
        pitch = np.arctan2(-R[2, 0], sy)    # Rotation around Y-axis
        roll = np.arctan2(R[1, 0], R[0, 0]) # Rotation around X-axis
    else:
        yaw = np.arctan2(-R[1, 2], R[1, 1])
        pitch = np.arctan2(-R[2, 0], sy)
        roll = 0

    # Convert radians to degrees
    return np.degrees([yaw, pitch, roll])

cap = cv2.VideoCapture(0)

prev_time = 0
new_time = 0

dim = 175
p1 = np.array([
    [0, 0],
    [dim - 1, 0],
    [dim - 1, dim - 1],
    [0, dim - 1]], dtype="float32")

while(True):
    ret, frame = cap.read()
    if ret == True:
        final_contour_list = contour_generator(frame)
        for i in range(len(final_contour_list)):
            x, y, w, h = cv2.boundingRect(final_contour_list[i])
            c_rez = final_contour_list[i][:, 0]
            h, _ = cv2.findHomography(order(c_rez), p1, cv2.RANSAC, 2)

            if h is not None:
                tag = cv2.warpPerspective(frame, h, (175, 175))
                markerColor = determineColor(tag)
                markerLetter = determineLetter(cv2.cvtColor(tag, cv2.COLOR_BGR2GRAY))
                tvec, angles = findTranslationAndRotation(order(c_rez)) # error here
                print("Tag: " + markerLetter + ", Color: " + markerColor)
                # TODO find the distance of the camera from the tag
                # TODO determine the coordinate of the specific tag by looking it up through a hard coded map of each tag's map coordinates
                # TODO determine the angle from the mid point at which the tag exists
                # TODO create an algorithm that uses distance, angle, and tag coordinates to determine position
                # TODO update low-level positioning data

                cv2.drawContours(frame, [final_contour_list[i]], -1, (0, 255, 0), 2)
                cv2.putText(frame, text=str(cv2.contourArea(final_contour_list[i])), org=(x, y-5), fontFace=cv2.FONT_HERSHEY_SIMPLEX, fontScale=0.25, color=(0, 255, 0), thickness=1, lineType=cv2.LINE_AA)
                cv2.imshow("Marker", tag)

        new_time = time.time()
        fps = 1/(new_time - prev_time)
        prev_time = new_time
        fps = int(fps)
        fps = str(fps)
        cv2.putText(frame, fps, (7, 70), cv2.FONT_HERSHEY_SIMPLEX, 2, (1000, 255, 0), 3, cv2.LINE_AA)
        cv2.imshow("Outline", frame)

    if cv2.waitKey(1) & 0xFF == ord('q'):
        break

# When everything done, release the capture
cap.release()
cv2.destroyAllWindows()

File: tag_detector/tag_detector.cpp
// tag_detect_pose.cpp
// Build (Ubuntu):
//   g++ -std=c++17 tag_detect_pose.cpp -o tag_detect `pkg-config --cflags --libs opencv4`
// Run:
//   ./tag_detect

#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <chrono>
#include <cmath>
#include <numeric>

using namespace cv;
using std::cout;
using std::endl;

// ---------------- Camera intrinsics (match your Python) ----------------
static const float fid_size_m = 0.053f; // meters
static const Mat cameraMatrix = (Mat_<float>(3,3) <<
    1.25649815e+03f, 0.f, 7.12996774e+02f,
    0.f, 1.25820533e+03f, 4.69551858e+02f,
    0.f, 0.f, 1.f);
static const Mat distCoeffs = (Mat_<float>(1,5) <<
    -3.72271817e-03f, 5.33786890e-01f, -4.99625728e-04f, -1.65101232e-03f, -1.78505927e+00f);

// ---------------- Color ranges (HSV) ----------------
static const Scalar lowerRed1(  0,150,150), upperRed1( 10,255,255);
static const Scalar lowerRed2(170,150,150), upperRed2(180,255,255);
static const Scalar lowerBlue(100, 80,160), upperBlue(120,255,255);

// ---------------- Helpers ----------------
static std::vector<std::vector<Point>> contour_generator(const Mat& frameBGR) {
    Mat gray, blurImg, edges;
    cvtColor(frameBGR, gray, COLOR_BGR2GRAY);
    GaussianBlur(gray, blurImg, Size(7,7), 0);

    // Median-based Canny thresholds (like Python)
    double med = medianBlur; // placeholder to avoid unused – we compute below
    {
        Mat tmp;
        blurImg.reshape(0,1).copyTo(tmp);
        // Fast median using nth_element
        std::vector<uchar> v(tmp.begin<uchar>(), tmp.end<uchar>());
        size_t mid = v.size()/2;
        std::nth_element(v.begin(), v.begin()+mid, v.end());
        med = v[mid];
    }
    int lower = (int)std::max(0.0, 0.3 * med);
    int upper = (int)std::min(255.0, 1.1 * med);

    Canny(blurImg, edges, lower, upper);
    imshow("Edges", edges);

    // Find contours with hierarchy (like RETR_TREE)
    std::vector<std::vector<Point>> cnts;
    std::vector<Vec4i> hierarchy;
    findContours(edges.clone(), cnts, hierarchy, RETR_TREE, CHAIN_APPROX_SIMPLE);

    if (hierarchy.empty()) return {};

    // Original code used complex parent logic and then filtered.
    // We’ll reproduce the intent: gather quads with area/convex filter later.
    std::vector<std::vector<Point>> result;
    for (size_t i = 0; i < cnts.size(); ++i) {
        // Try approximating this contour and (optionally) its previous sibling, like Python did
        double peri = arcLength(cnts[i], true);
        std::vector<Point> approx; approxPolyDP(cnts[i], approx, 0.02 * peri, true);
        if (approx.size() > 4 && i > 0) {
            double peri1 = arcLength(cnts[i-1], true);
            std::vector<Point> corners; approxPolyDP(cnts[i-1], corners, 0.02 * peri1, true);
            result.push_back(corners);
        }
    }

    // Keep only quads, area>1000, convex
    std::vector<std::vector<Point>> final_list;
    for (auto& contour : result) {
        if (contour.size() == 4 && contourArea(contour) > 1000.0 && isContourConvex(contour)) {
            final_list.push_back(contour);
        }
    }
    return final_list;
}

static std::vector<Point2f> order(const std::vector<Point2f>& pts) {
    CV_Assert(pts.size() == 4);
    std::vector<Point2f> rect(4);
    std::vector<float> s(4), d(4);
    for (int i=0;i<4;++i){ s[i]=pts[i].x+pts[i].y; d[i]=pts[i].y-pts[i].x; }
    int tl = int(std::distance(s.begin(), std::min_element(s.begin(), s.end())));
    int br = int(std::distance(s.begin(), std::max_element(s.begin(), s.end())));
    int tr = int(std::distance(d.begin(), std::min_element(d.begin(), d.end())));
    int bl = int(std::distance(d.begin(), std::max_element(d.begin(), d.end())));
    rect[0]=pts[tl]; rect[2]=pts[br]; rect[1]=pts[tr]; rect[3]=pts[bl];
    return rect;
}

// Determine main color inside 175x175 tag image (BGR)
static std::string determineColor(const Mat& markerBGR) {
    Mat hsv; cvtColor(markerBGR, hsv, COLOR_BGR2HSV);
    Mat red1, red2, red, blue;
    inRange(hsv, lowerRed1, upperRed1, red1);
    inRange(hsv, lowerRed2, upperRed2, red2);
    red = red1 + red2;
    inRange(hsv, lowerBlue, upperBlue, blue);

    imshow("redMask1", red1);
    imshow("redMask2", red2);
    imshow("redMask", red);
    imshow("blueMask", blue);

    if (sum(red)[0] > 0)  return "red";
    if (sum(blue)[0] > 0) return "blue";
    return {};
}

// Determine letter from warped grayscale tag
static std::string determineLetter(const Mat& markerGray) {
    Mat img_bw; threshold(markerGray, img_bw, 200, 255, THRESH_BINARY);
    const uchar white = 255;

    if (img_bw.at<uchar>(5,5) == white) {
        return {}; // False
    }
    Mat cropped = img_bw(Rect(25,25,125,125));
    imshow("Cropped", cropped);

    auto px = [&](int y, int x)->uchar { return cropped.at<uchar>(y,x); };

    // replicate your pixel tests (note: second comment said 'B' but code returned 'C')
    if (px(12,37) != white)                              return "A";
    else if (px(12,12) != white)                         return "C";
    else if (px(12,112) == white)                        return "E";
    else if (px(62,112) == white)                        return "D";
    else                                                 return "B";
}

static cv::Vec3d rotToEul(const Mat& R) {
    double sy = std::sqrt(R.at<double>(0,0)*R.at<double>(0,0) + R.at<double>(1,0)*R.at<double>(1,0));
    bool singular = sy < 1e-6;
    double yaw, pitch, roll;
    if (!singular) {
        yaw   = std::atan2(R.at<double>(2,1), R.at<double>(2,2));
        pitch = std::atan2(-R.at<double>(2,0), sy);
        roll  = std::atan2(R.at<double>(1,0), R.at<double>(0,0));
    } else {
        yaw   = std::atan2(-R.at<double>(1,2), R.at<double>(1,1));
        pitch = std::atan2(-R.at<double>(2,0), sy);
        roll  = 0.0;
    }
    return { yaw*180.0/M_PI, pitch*180.0/M_PI, roll*180.0/M_PI };
}

static bool findTranslationAndRotation(const std::vector<Point2f>& imgPts,
                                       Vec3d& tvec, Vec3d& rpy_deg)
{
    std::vector<Point3f> objPts = {
        { -fid_size_m/2.f,  fid_size_m/2.f, 0.f },
        {  fid_size_m/2.f,  fid_size_m/2.f, 0.f },
        {  fid_size_m/2.f, -fid_size_m/2.f, 0.f },
        { -fid_size_m/2.f, -fid_size_m/2.f, 0.f }
    };
    Mat rvec, tvecMat;
    bool ok = solvePnP(objPts, imgPts, cameraMatrix, distCoeffs, rvec, tvecMat, false, SOLVEPNP_IPPE_SQUARE);
    if (!ok) return false;

    Mat R; Rodrigues(rvec, R);
    rpy_deg = rotToEul(R);
    tvec = Vec3d(tvecMat.at<double>(0,0), tvecMat.at<double>(1,0), tvecMat.at<double>(2,0));
    return true;
}

int main() {
    VideoCapture cap(0);
    if (!cap.isOpened()) {
        std::cerr << "Error: Could not open webcam.\n";
        return 1;
    }

    const int dim = 175;
    std::vector<Point2f> p1 = {
        {0.f,0.f}, {(float)(dim-1),0.f}, {(float)(dim-1),(float)(dim-1)}, {0.f,(float)(dim-1)}
    };

    auto t_prev = std::chrono::steady_clock::now();

    Mat frame;
    while (true) {
        if (!cap.read(frame) || frame.empty()) break;

        auto contours = contour_generator(frame);

        for (const auto& c : contours) {
            // draw detected contour
            polylines(frame, c, true, Scalar(0,255,0), 2);

            // order corners
            std::vector<Point2f> corners2f;
            for (auto& p : c) corners2f.emplace_back((float)p.x, (float)p.y);
            auto ord = order(corners2f);

            // homography
            Mat H = findHomography(ord, p1, RANSAC, 2.0);
            if (H.empty()) continue;

            // warp
            Mat tag; warpPerspective(frame, tag, H, Size(dim, dim));
            std::string letter = determineLetter(cvtColor(tag, tag, COLOR_BGR2GRAY), ""); // placeholder to match overload
            // The above line is incorrect; fix: separate gray conversion
            {
                Mat grayTag; cvtColor(tag, grayTag, COLOR_BGR2GRAY);
                letter = determineLetter(grayTag);
            }
            std::string color = determineColor(tag);
            if (letter.empty() || color.empty()) continue;

            Vec3d tvec, rpy;
            if (findTranslationAndRotation(ord, tvec, rpy)) {
                cout << "Tag: " << letter << ", Color: " << color << endl;
                // Match your print mapping: x=tvec[2], y=tvec[0], z=tvec[1]
                cout << "CAMERA Position x=" << tvec[2] << " y=" << tvec[0] << " z=" << tvec[1] << endl;
                cout << "Camera tilt -------   yaw=" << rpy[0]
                     << " pitch=" << rpy[1]
                     << " roll="  << rpy[2] << endl;

                // area debug
                Rect bbox = boundingRect(c);
                putText(frame, std::to_string((int)contourArea(c)), {bbox.x, bbox.y-5},
                        FONT_HERSHEY_SIMPLEX, 0.5, Scalar(0,255,0), 1, LINE_AA);

                imshow("Marker", tag);
            }
        }

        // FPS overlay
        auto t_now = std::chrono::steady_clock::now();
        double dt = std::chrono::duration<double>(t_now - t_prev).count();
        t_prev = t_now;
        int fps = (dt > 0.0) ? (int)std::round(1.0 / dt) : 0;
        putText(frame, std::to_string(fps), {7,70}, FONT_HERSHEY_SIMPLEX, 2.0, Scalar(0,255,0), 3, LINE_AA);

        imshow("Outline", frame);
        if ((waitKey(1) & 0xFF) == 'q') break;
    }

    cap.release();
    destroyAllWindows();
    return 0;
}

